[NNC] Added matmul for NNC lowering/unified dtypes

Chillee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56718 Added skeleton of python dispatch key
* #56775 [NNC] ported over sum and softmax to new scheme
* #56766 [nnc] ported some more ops + added vectors to argvalue
* #56679 moved lowerings out of the TensorExprKernel and into independent functions
* **#56456 [NNC] Added matmul for NNC lowering/unified dtypes**



Differential Revision: [D27977532](https://our.internmc.facebook.com/intern/diff/D27977532)

Chillee committed:
Added matmul/unified dtypes

[ghstack-poisoned]

Chillee committed:
Update on "Added matmul/unified dtypes"




[ghstack-poisoned]

Chillee committed:
Update on "Added matmul/unified dtypes"




[ghstack-poisoned]

Chillee committed:
Update on "Added matmul/unified dtypes"




[ghstack-poisoned]

Chillee committed:
Update on "Added matmul/unified dtypes"




[ghstack-poisoned]

ZolotukhinM reviewed (approved):
Overall looks good! Please include a more detailed description in the commit message and add a tag ([NNC] or [TensorExpr]) - it helps to find relevant commits for release notes.

Chillee committed:
Update on "[NNC] Added matmul for NNC lowering/unified dtypes"




[ghstack-poisoned]

Chillee committed:
Update on "[NNC] Added matmul for NNC lowering/unified dtypes"




[ghstack-poisoned]

Chillee committed:
Update on "[NNC] Added matmul for NNC lowering/unified dtypes"




[ghstack-poisoned]

Chillee committed:
Update on "[NNC] Added matmul for NNC lowering/unified dtypes"




Differential Revision: [D27977532](https://our.internmc.facebook.com/intern/diff/D27977532)

[ghstack-poisoned]
